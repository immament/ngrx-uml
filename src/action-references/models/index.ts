export * from './action-reference.model';
