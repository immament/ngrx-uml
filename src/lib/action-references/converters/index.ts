export { ActionReferenceConverter } from './node-converters/action-reference.converter';

