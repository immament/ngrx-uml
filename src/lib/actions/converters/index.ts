export * from './action-convert.context';
export * from './node-converters';
