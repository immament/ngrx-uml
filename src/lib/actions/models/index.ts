export * from './action.model';
export * from './action-with-symbol.model';
export * from './plant-item.model';
