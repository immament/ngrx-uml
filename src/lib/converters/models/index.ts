export { NodeConverter } from './node.converter';
export * from './type.model';
