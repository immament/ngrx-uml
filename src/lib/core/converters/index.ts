export * from './convert.context';
export * from './converter';
