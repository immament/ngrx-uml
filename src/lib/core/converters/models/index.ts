// start:ng42.barrel
export * from './converted-items/call-expression.model';
export * from './converted-items/named-converted-item.model';
export * from './converted-items/named-type.model';
export * from './property.model';
export * from './type-kind.enum';
export * from './converted-items/type-argument.model';
export * from './converted-items/converted-item.model';
export * from './type-literal.model';
export * from './type-reference.model';
// end:ng42.barrel

