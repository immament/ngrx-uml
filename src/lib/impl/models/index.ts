// start:ng42.barrel
export * from './action-reference.model';
export * from './item-with-symbol.model';
export * from './action.model';
export * from './reducer.model';
// end:ng42.barrel

