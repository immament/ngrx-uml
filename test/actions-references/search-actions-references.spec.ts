
import log from 'loglevel';
import path from 'path';
import ts from 'typescript';

import { ConvertContext, ConvertContextFactory, Converter } from '../../src/lib/core/converters';
import { NamedConvertedItem, TypeKind } from '../../src/lib/core/converters/models';
import {
    ActionConvertContextFactory,
} from '../../src/lib/impl/converters/action-convert-context.factory';
import {
    ActionReferenceConvertContextFactory,
} from '../../src/lib/impl/converters/action-reference-convert-context.factory';
import { Action, ActionReference } from '../../src/lib/impl/models';

describe('Search Actions References', () => {
    let actionConvertFactory: ActionConvertContextFactory;
    let actionReferenceFactory: ActionReferenceConvertContextFactory;

    beforeAll(() => {
        log.setLevel(log.levels.WARN);
        actionConvertFactory = new ActionConvertContextFactory;
        actionReferenceFactory = new ActionReferenceConvertContextFactory;
        log.trace = log.debug;
    });

    const base = path.join(__dirname, 'search-actions-references_data');

    function createPathToTestFile(fileName: string): string {
        return path.join(base, fileName);
    }


    function convert(program: ts.Program, convertFactories: ConvertContextFactory[]): Map<TypeKind, NamedConvertedItem[]> | undefined {
        const converter = new Converter();
        const typeChecker = program.getTypeChecker();
        let converterResult: Map<TypeKind, NamedConvertedItem[]> | undefined;
        let lastContext: ConvertContext | undefined;
        for (const contextFactory of convertFactories) {
            const context = contextFactory.create(program, typeChecker, converter, lastContext);
            converterResult = converter.convert(context, program);
            lastContext = context;
        }
        return converterResult;
    }

    /**
     * Remove file paths diffrent in enviroments
     */
    function prepareActionReferencesToSnapshot(actionReferences: ActionReference[]): void {
        actionReferences?.forEach(ar => {
            ar.fileName = undefined;
            if (ar.action) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (ar.action as any).filePath = undefined;
            }
        });

    }

    function checkResult(result: Map<TypeKind, NamedConvertedItem[]> | undefined): void {
        expect(result).toBeDefined();

        const actions = result?.get(TypeKind.Action);
        expect(actions).toBeDefined();
        expect(actions).toHaveLength(1);
        if (!actions) return;
        const actionReferences = (actions[0] as Action).references;
        expect(actionReferences).toBeDefined();
        expect(actionReferences).toHaveLength(1);
        if (actionReferences) {
            prepareActionReferencesToSnapshot(actionReferences);
            expect(actionReferences).toMatchSnapshot();

        }
    }

    it('Should find one action reference WITH alias', () => {
        const programFiles = ['references/with-alias.ts', 'actions/actions.ts'];
        const program = ts.createProgram([...programFiles.map((fileName) => createPathToTestFile(fileName))], {});
        const result = convert(program, [actionConvertFactory, actionReferenceFactory]);
        checkResult(result);
    });

    it('Should find one action reference WITHOUT alias', () => {
        const programFiles = ['references/without-alias.ts', 'actions/actions.ts'];
        const program = ts.createProgram([...programFiles.map((fileName) => createPathToTestFile(fileName))], {});
        const result = convert(program, [actionConvertFactory, actionReferenceFactory]);
        checkResult(result);
    });

    it('Should find one action reference WITH tsconfig path and import alias', () => {
        const programFiles = ['references/with-tsconfig-path-with-import-alias.ts', 'actions/actions.ts'];
        const program = ts.createProgram([...programFiles.map((fileName) => createPathToTestFile(fileName))], {
            baseUrl: base,
            paths: {
                '@app/*': ['./actions/*']
            }
        });

        log.info('program.getCompilerOptions()', program.getCompilerOptions());
        const result = convert(program, [actionConvertFactory, actionReferenceFactory]);
        checkResult(result);
    });

    it('Should find one action reference WITH tsconfig path and WITHOUT import alias', () => {
        const programFiles = ['references/with-tsconfig-path-without-import-alias.ts', 'actions/actions.ts'];
        const program = ts.createProgram([...programFiles.map((fileName) => createPathToTestFile(fileName))], {
            baseUrl: base,
            paths: {
                '@app/*': ['./actions/*']
            }
        });

        log.info('program.getCompilerOptions()', program.getCompilerOptions());
        const result = convert(program, [actionConvertFactory, actionReferenceFactory]);
        checkResult(result);
    });
});