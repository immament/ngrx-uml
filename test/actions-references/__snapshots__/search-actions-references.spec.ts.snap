// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Actions References Should find one action reference WITH alias 1`] = `
Array [
  ActionReference {
    "action": Action {
      "createActionArgs": undefined,
      "end": 83,
      "filePath": undefined,
      "kind": 1002,
      "kindText": "Action",
      "name": "[Books] Remove Book",
      "pos": 42,
      "references": [Circular],
      "variable": "removeBook",
    },
    "declarationContext": Array [],
    "documentation": "",
    "fileName": "with-alias",
    "filePath": undefined,
    "isCall": true,
    "kind": 1003,
    "kindText": "ActionReference",
    "name": "removeBook",
    "type": "any",
  },
]
`;

exports[`Search Actions References Should find one action reference WITH tsconfig path and WITHOUT import alias 1`] = `
Array [
  ActionReference {
    "action": Action {
      "createActionArgs": undefined,
      "end": 83,
      "filePath": undefined,
      "kind": 1002,
      "kindText": "Action",
      "name": "[Books] Remove Book",
      "pos": 42,
      "references": [Circular],
      "variable": "removeBook",
    },
    "declarationContext": Array [],
    "documentation": "",
    "fileName": "with-tsconfig-path-without-import-alias",
    "filePath": undefined,
    "isCall": true,
    "kind": 1003,
    "kindText": "ActionReference",
    "name": "removeBook",
    "type": "any",
  },
]
`;

exports[`Search Actions References Should find one action reference WITH tsconfig path and import alias 1`] = `
Array [
  ActionReference {
    "action": Action {
      "createActionArgs": undefined,
      "end": 83,
      "filePath": undefined,
      "kind": 1002,
      "kindText": "Action",
      "name": "[Books] Remove Book",
      "pos": 42,
      "references": [Circular],
      "variable": "removeBook",
    },
    "declarationContext": Array [],
    "documentation": "",
    "fileName": "with-tsconfig-path-with-import-alias",
    "filePath": undefined,
    "isCall": true,
    "kind": 1003,
    "kindText": "ActionReference",
    "name": "removeBook",
    "type": "any",
  },
]
`;

exports[`Search Actions References Should find one action reference WITHOUT alias 1`] = `
Array [
  ActionReference {
    "action": Action {
      "createActionArgs": undefined,
      "end": 83,
      "filePath": undefined,
      "kind": 1002,
      "kindText": "Action",
      "name": "[Books] Remove Book",
      "pos": 42,
      "references": [Circular],
      "variable": "removeBook",
    },
    "declarationContext": Array [],
    "documentation": "",
    "fileName": "without-alias",
    "filePath": undefined,
    "isCall": true,
    "kind": 1003,
    "kindText": "ActionReference",
    "name": "removeBook",
    "type": "any",
  },
]
`;
